# Analyze placementdata - 10k.csv
Clean dataset to have numbers and normalize each column

In [2]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
datasetFileName = "../data/placementdata - 10k.csv"

placementDataFrame = pd.read_csv(datasetFileName)

placementDataFrame.head()


,StudentID,CGPA,Internships,Projects,Workshops/Certifications,AptitudeTestScore,SoftSkillsRating,ExtracurricularActivities,PlacementTraining,SSC_Marks,HSC_Marks,PlacementStatus
0,1,7.5,1,1,1,65,4.4,No,No,61,79,NotPlaced
1,2,8.9,0,3,2,90,4.0,Yes,Yes,78,82,Placed
2,3,7.3,1,2,2,82,4.8,Yes,No,79,80,NotPlaced
3,4,7.5,1,1,2,85,4.4,Yes,Yes,81,80,Placed
4,5,8.3,1,2,2,86,4.5,Yes,Yes,74,88,Placed


# Data Analysis
## Averages:
- CGPA - 7.69800999999997
- Internships - 1.0492
- Projects - 2.0266
- Workshops/Certifications - 1.0132
- AptitudeTestScore - 79.4499
- SoftSkillsRating - 4.3239600000000395
- SSC_Marks - 69.1594
- HSC_Marks - 74.5015

## Percentages
### ExtracurricularActivities
 - No - 41.46  Yes - 58.540000000000006  
### PlacementTraining
 - No - 26.82  Yes - 73.18  
### PlacementStatus
 - NotPlaced - 58.03  Placed - 41.97 

In [13]:
# Drop StudentID (not a predictive feature)
X = placementDataFrame.drop(columns=["StudentID", "PlacementStatus", "AptitudeTestScore", "Workshops/Certifications", ])
y = placementDataFrame["PlacementStatus"]

X.head()

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and train model
model = LogisticRegression()
model.fit(X_train, y_train)

# Predict and evaluate
y_pred = model.predict(X_test)

# print("Accuracy:", accuracy_score(y_test, y_pred))
# print(classification_report(y_test, y_pred))

# Get probability predictions
y_proba = model.predict_proba(X_test)

# y_proba is a 2D array where:
# Column 0 = probability of class 0 (Not Placed)
# Column 1 = probability of class 1 (Placed)

print("First 5 predictions with probabilities:")
for i in range(20):
    prob_class_0 = y_proba[i][0] * 100
    prob_class_1 = y_proba[i][1] * 100
    predicted_class = y_pred[i]
    actual_class = y_test.iloc[i]
    
    print(f"Sample {i+1}:")
    print(f"  Probability of Not Placed (0): {prob_class_0:.2f}%")
    print(f"  Probability of Placed (1): {prob_class_1:.2f}%")
    print(f"  Predicted: {predicted_class}, Actual: {actual_class}\n")


First 5 predictions with probabilities:
Sample 1:
  Probability of Not Placed (0): 95.85%
  Probability of Placed (1): 4.15%
  Predicted: 0, Actual: 0

Sample 2:
  Probability of Not Placed (0): 90.40%
  Probability of Placed (1): 9.60%
  Predicted: 0, Actual: 0

Sample 3:
  Probability of Not Placed (0): 98.12%
  Probability of Placed (1): 1.88%
  Predicted: 0, Actual: 0

Sample 4:
  Probability of Not Placed (0): 8.17%
  Probability of Placed (1): 91.83%
  Predicted: 1, Actual: 0

Sample 5:
  Probability of Not Placed (0): 35.72%
  Probability of Placed (1): 64.28%
  Predicted: 1, Actual: 1

Sample 6:
  Probability of Not Placed (0): 11.61%
  Probability of Placed (1): 88.39%
  Predicted: 1, Actual: 1

Sample 7:
  Probability of Not Placed (0): 25.26%
  Probability of Placed (1): 74.74%
  Predicted: 1, Actual: 1

Sample 8:
  Probability of Not Placed (0): 95.82%
  Probability of Placed (1): 4.18%
  Predicted: 0, Actual: 0

Sample 9:
  Probability of Not Placed (0): 75.21%
  Probabili

## Different Models Comparison for Placement Data

### Logistic Regression Model
- 79% accuracy with Logistic Regression Model excluding features which are dropped

---

# Analyse student career suggestion - 20k.csv

In [3]:
import pandas as pd
# from collections import Counter

suggestionDataFileName = "../data/student career suggestion - 20k.csv"

suggestionDF = pd.read_csv(suggestionDataFileName)

suggestionDF.head()


,Acedamic percentage in Operating Systems,percentage in Algorithms,Percentage in Programming Concepts,Percentage in Software Engineering,Percentage in Computer Networks,Percentage in Electronics Subjects,Percentage in Computer Architecture,Percentage in Mathematics,Percentage in Communication skills,Hours working per day,...,Interested Type of Books,Salary Range Expected,In a Realtionship?,Gentle or Tuff behaviour?,Management or Technical,Salary/work,hard/smart worker,worked in teams ever?,Introvert,Suggested Job Role
0,69,63,78,87,94,94,87,84,61,9,...,Prayer books,salary,no,stubborn,Management,salary,hard worker,yes,no,Database Developer
1,78,62,73,60,71,70,73,84,91,12,...,Childrens,salary,yes,gentle,Technical,salary,hard worker,no,yes,Portal Administrator
2,71,86,91,87,61,81,72,72,94,11,...,Travel,Work,no,stubborn,Management,work,hard worker,no,yes,Portal Administrator
3,76,87,60,84,89,73,62,88,69,7,...,Romance,Work,yes,gentle,Management,work,smart worker,yes,yes,Systems Security Administrator
4,92,62,90,67,71,89,73,71,73,4,...,Cookbooks,salary,no,stubborn,Management,work,hard worker,yes,yes,Business Systems Analyst


In [5]:
# # Random Forest for feature importance

# from sklearn.model_selection import train_test_split
# from sklearn.ensemble import RandomForestClassifier

# y = suggestionDF["Suggested Job Role"]
# X = suggestionDF.drop(columns=["Suggested Job Role"])

# # one-hot encoding
# X = pd.get_dummies(X)


# X_train, X_test, y_train, y_test = train_test_split(
#     X, y, test_size=0.2, random_state=42
# )

# rf = RandomForestClassifier(
#     n_estimators=300,
#     random_state=42,
#     n_jobs=-1
# )


# rf.fit(X_train, y_train)

# # Extract importances
# importances = rf.feature_importances_
# features = X.columns

# # Put them in a DataFrame for convenience
# importance_df = pd.DataFrame({
#     "feature": features,
#     "importance": importances
# }).sort_values(by="importance", ascending=False)

# importance_df["group"] = importance_df["feature"].str.rsplit("_", 1).str[0]

# grouped_importance = (importance_df
#                       .groupby("group")["importance"]
#                       .sum()
#                       .sort_values(ascending=False))

# grouped_importance

# Dataset Stats (student career suggestion - 20k.csv)

34 total unique job roles
- Database Developer -> 581
- Portal Administrator -> 593
- Systems Security Administrator -> 562
- Business Systems Analyst -> 582
- Software Systems Engineer -> 575
- Business Intelligence Analyst -> 540
- CRM Technical Developer -> 567
- Mobile Applications Developer -> 538
- UX Designer -> 589
- Quality Assurance Associate -> 565
- Web Developer -> 570
- Information Security Analyst -> 543
- CRM Business Analyst -> 584
- Technical Support -> 565
- Project Manager -> 602
- Information Technology Manager -> 591
- Programmer Analyst -> 529
- Design & UX -> 588
- Solutions Architect -> 578
- Systems Analyst -> 550
- Network Security Administrator -> 1112
- Data Architect -> 564
- Software Developer -> 587
- E-Commerce Analyst -> 546
- Technical Services/Help Desk/Tech Support -> 558
- Information Technology Auditor -> 558
- Database Manager -> 570
- Applications Developer -> 551
- Database Administrator -> 593
- Network Engineer -> 621
- Software Engineer -> 590
- Technical Engineer -> 557
- Network Security Engineer -> 630
- Software Quality Assurance (QA) / Testing -> 571


# Feature Importance based on Random Forest
- 3	Percentage in Software Engineering	0.032201
- 0	Acedamic percentage in Operating Systems	0.032135
- 8	Percentage in Communication skills	0.032107
- 6	Percentage in Computer Architecture	0.032056
- 1	percentage in Algorithms	0.031993
- ...	...	...
- 83	Interested Type of Books_Biographies	0.001996
- 109	Interested Type of Books_Trilogy	0.001995
- 88	Interested Type of Books_Dictionaries	0.001978
- 97	Interested Type of Books_Math	0.001923
- 87	Interested Type of Books_Diaries	0.001905


In [7]:
# Transform dataset

columns_to_cgpa = suggestionDF.columns[0:8] #first 7 columns for cgpa
no_of_subjects = len(columns_to_cgpa)
# print(columns_to_cgpa)

suggestionDF[columns_to_cgpa] = suggestionDF[columns_to_cgpa].astype(float) / 10

suggestionDF["Percentage in Communication skills"] = suggestionDF["Percentage in Communication skills"].astype(float)/10

suggestionDF["CombinedScore"] = suggestionDF[columns_to_cgpa].mean(axis=1)

suggestionDF[["CombinedScore"]]

cols_order = [suggestionDF.columns[-1]] + list(suggestionDF.columns[8:-1])

suggestionDF = suggestionDF[cols_order]

suggestionDF.head()

suggestionDF.to_csv("../data/career suggestion - transformed.csv", index=False)

In [2]:
# Random Forest for feature importance

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
import pandas as pd

suggestionDFTransformed = pd.read_csv("../data/career suggestion - transformed.csv")

y = suggestionDFTransformed["Suggested Job Role"]
X = suggestionDFTransformed.drop(columns=["Suggested Job Role"])

# one-hot encoding
X = pd.get_dummies(X)


X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

rf = RandomForestClassifier(
    n_estimators=300,
    random_state=42,
    n_jobs=-1
)


rf.fit(X_train, y_train)

# Extract importances
importances = rf.feature_importances_
features = X.columns

# Put them in a DataFrame for convenience
importance_df = pd.DataFrame({
    "feature": features,
    "importance": importances
}).sort_values(by="importance", ascending=False)

importance_df

,feature,importance
0,CombinedScore,0.043817
1,Percentage in Communication skills,0.040374
6,public speaking points,0.031785
3,Logical quotient rating,0.031762
2,Hours working per day,0.031665
...,...,...
76,Interested Type of Books_Biographies,0.002614
81,Interested Type of Books_Dictionaries,0.002614
79,Interested Type of Books_Cookbooks,0.002568
90,Interested Type of Books_Math,0.002504


# Different Models for Job Role Suggestion
34 classifications with 126 features
- Random Forest - 4.2% Accuracy
- Logistic Regression - 3% Accuracy (limit reached on 5000 iterations)

Dataset refinement needed (or not much correlation to the features)